### PR TITLE
Exercise 3

### PR DESCRIPTION
**Pull request by achamberlain999**
Features added:

- Unit tests for homepage view model
- Integration test for homepage endpoint

Potential improvements:

- Improve test coverage
- End-to-end tests
- Improve code structure: I've tried splitting out some classes but I find it's difficult to know what a good codebase structure looks like without experience!

Submission

- [x] Submit to Aptem
- [x] Finish clicked

### CODE REVIEW

File: todo_app/network/trello_client.py [hunk 1]
```diff
@@ -1,60 +1,62 @@
...
+    def __init__(self, board_id, trello_api_base_url, trello_api_key, trello_api_token):
+        self.BOARD_ID = board_id
+        self.TRELLO_API_BASE_URL = trello_api_base_url
+        self.TRELLO_API_KEY = trello_api_key
+        self.TRELLO_API_TOKEN = trello_api_token
+        self.LIST_IDS = {
+            'To do': None,
+            'Done': None
+        }
+        self.get_list_ids()
```
Comment (CorndelDevOps) [inline todo_app/network/trello_client.py]: After a bit of digging, I think this is one of the reasons why the integration test is failing. You're making a trello request before the stubbing logic is called, so in here you're not mocking the request, you're actually doing a request to trello.

I think the easiest way of solving this would be to remove it from the constructor, and call it in the methods below, when necessary, in the form of e.g.:
```
if self.LIST_IDS['To Do'] is None or self.LIST_IDS['Done'] is None:
    self.get_list_ids()
```

File: todo_app/tests/integration/test_client.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        assert "Test card" in response.data.decode()
+
+class StubResponse():
+    def __init__(self, fake_response_data):
+        self.fake_response_data = fake_response_data
+
+    def json(self):
+        return self.fake_response_data
+
+def get_lists_stub(url, params):
```
Comment (CorndelDevOps) [inline todo_app/tests/integration/test_client.py]: I'm not entirely sure why, but to get the test running locally I needed to add a default value to params, i.e. `params=None`

File: env/.env.test [hunk 3]
```diff
@@ -0,0 +1,8 @@
+# Flask server configuration.
+FLASK_APP=todo_app/app
+FLASK_ENV=development
+
+# Change the following values for local development.
+SECRET_KEY=test_secret_key
+
+TRELLO_BOARD_ID=test_board_id
```
Comment (CorndelDevOps) [inline env/.env.test:12]: It looks like `TRELLO_API_BASE_URL` is missing from this file. Also, when you're testing, the path that you use is `.env.test`, but this file is now in `env/.env.test`

**Review by CorndelDevOps — COMMENTED** [review]
You're nearly there. There are a couple of things that I needed to do to get the integration test running, but otherwise looks good to me :)

File: todo_app/network/trello_client.py (hunk 1 above)
Comment (achamberlain999) [reply todo_app/network/trello_client.py]: Whoops! Thank you, difficult to spot an issue like this when you have the API keys set up locally so the request works

File: todo_app/templates/index.html [hunk 4]
```diff
@@ -18,16 +18,16 @@ <h2 class="text-center" style="margin: 40px 0 20px">Tasks</h2>
     <div class="col-lg">
       <h5 class="text-center">Open</h5>
       <ul class="list-group mb-4">
-      {% if not uncompleted_items %}
-      {% if completed_items %}
+      {% if not model.uncompleted_items %}
+      {% if model.completed_items %}
```
Comment (BogRad97) [inline todo_app/templates/index.html:22]: It's a bit hard to follow if the blocks are not indented, but dw, you don't need to fix it

**Review by BogRad97 — COMMENTED** [review]
Looks good to me :)
